code/iaas/agent-instance/src/main/java/io/cattle/platform/agent/instance/dao/impl/AgentInstanceDaoImpl.java

```diff
@@ -95,11 +97,16 @@ public boolean areAllCredentialsActive(Agent agent) {
                     .on(INSTANCE_LABEL_MAP.INSTANCE_ID.eq(INSTANCE.ID))
                 .join(LABEL)
                     .on(LABEL.ID.eq(INSTANCE_LABEL_MAP.LABEL_ID).and(LABEL.KEY.eq(providedServiceLabel)))
+                .join(INSTANCE_HOST_MAP)
+                    .on(INSTANCE.ID.eq(INSTANCE_HOST_MAP.INSTANCE_ID))
+                .join(HOST)
+                    .on(INSTANCE_HOST_MAP.HOST_ID.eq(HOST.ID))
                 .where(INSTANCE.ACCOUNT_ID.eq(accountId)
                     .and(INSTANCE.AGENT_ID.isNotNull())
                         .and(INSTANCE.STATE.eq(InstanceConstants.STATE_RUNNING))
                         .and(INSTANCE.HEALTH_STATE.in(HealthcheckConstants.HEALTH_STATE_HEALTHY,
                                 HealthcheckConstants.HEALTH_STATE_UPDATING_HEALTHY)))
+                        .and(HOST.AGENT_STATE.isNull().or(HOST.AGENT_STATE.notIn(AgentConstants.STATE_DISCONNECTED, AgentConstants.STATE_DISCONNECTING, AgentConstants.STATE_RECONNECTING)))
```

Comment:
host.agent_state is not good enough as it is not entirely accurate. Also, we just want to target state=active. Here's what we need overall:
```
select
...
join instance_host_map
join host
join agent on host.agent_id = id
where
agent.state == 'active'
```

Can you also add a test that exercises this? the test can be in test_resource_based_scheduler and should do the folowing:
1. Use a new_context
2. Add a container to that context that will register itself as an external scheduler using this logic: https://github.com/rancher/cattle/blob/d35dcbebb23d99b4b619ad0ff2955b0e2f7102d2/tests/integration/cattletest/core/test_resource_based_scheduling.py#L218-L226
3. Delete the host that the container was deployed on
4. Add a new host to the context
5. Ensure you can deploy a container